### ISSUE TITLE
Face and edge color in points layer incorrect if not all values in properties

### ISSUE THREAD

**Issue by constantinpape** [napari/napari#1348]
## 🐛 Bug

The mapping of edge and face colors using `face_color_cycle` / `edge_color_cycle` is incorrect if 
any of the possible property values is missing from the current property values.
This is due to the use of np.unique to determine the current property values:
https://github.com/napari/napari/blob/c04484fc08694d23b4790387c3618d9fb4fad00f/napari/layers/points/points.py#L288
https://github.com/napari/napari/blob/c04484fc08694d23b4790387c3618d9fb4fad00f/napari/layers/points/points.py#L1037

## To Reproduce

Take this toy example, where I want to map points with value 1 to green, 2 to red, etc.
But where in the initial points the property 1 is not present.

```python
points = np.array([[0, 0], [1, 1], [2, 2]])

# property to colors that I want to achieve:
color_map_expected = {1: 'green', 2: 'red', 3: 'blue', 4: 'white'}

# how to currently do it in napari
properties = {'my_colors': [2, 3, 4]}
face_color_cycle = ['green', 'red', 'blue', 'white']
viewer.add_points(points, face_color='my_colors', properties=properties)
```
This will show the points in green, red and blue instead of red, blue, white as intended.

## Environment

Current napari master.

## Additional context

To take a real world example, this happens in my code if for example the "unlabeled" property (value = 0) is missing: 
https://github.com/hci-unihd/covid-if-annotations/blob/master/napari_covid_if_annotations/layers.py#L26-L44

In general, I find specifying the `color_cycles` via lists and then the mapping via `zip(np.unique(values), color_cycle)` not intuitive and think maps would make a lot more sense here.

**Comment by kevinyamauchi** [comment]
Hey @constantinpape ! Thanks for the submitting an issue.

I am sorry about the color cycle mapping behavior. At the moment, we do not have a way in the init method to specify the mapping between the property value and color (it works something like `zip(np.unique(values), color_cycle)` as you mentioned). I agree that the ability to explicitly map the property values and color for the color cycles in the init method is an important feature.

**A potential fix for now**
In the meantime, you can specify the mapping between property value and color via the properties`Points.face_color_cycle_map` or `Points.edge_color_cycle_map`. For example, using the sample code in your issue I think the following would work. This isn't really the "official" way to use the color cycle, but it should work for now while we figure out the best way to implement it in the init method.

```python
points = np.array([[0, 0], [1, 1], [2, 2]])

# property to colors that I want to achieve:
face_color_cycle_map = {
    1: [0, 1, 0, 1],
    2: [1, 0, 0, 1],
    3: [0, 0, 1, 1],
    4: [1, 1, 1, 1]
}

# how to currently do it in napari
properties = {'my_colors': [2, 3, 4]}
points_layer = viewer.add_points(points, face_color='my_colors', properties=properties)

# set the face color cycle map
points_layer.face_color_cycle_map = face_color_cycle_map

# we have to refresh the colors to update them
points_layer.refresh_colors()
```

Note that when using this method, the colors need to be RGBA numpy arrays and we have to refresh the colors after setting the `face_color_cycle_map`. However, once you've done that you should be able to add/toggle points as you normally would. The one caveat that adding points with property values not in 'face_color_cycle_map' would result in those new points getting colors assigned from the face color cycle (if not specified, defaults to `['magenta', 'green']`)

**Potential improvement for the API**
How would you prefer the API to be? Perhaps allowing the user to pass a dictionary to the color cycle to specify the mapping and a list to allow apply a color cycle to the unique properties (as it is now)?
```python
points = np.array([[0, 0], [1, 1], [2, 2]])

# make a points layer using a specified property value to color mapping
color_map_specified = {1: 'green', 2: 'red', 3: 'blue', 4: 'white'}
properties = {'my_colors': [2, 3, 4]}
viewer.add_points(points, face_color='my_colors', face_color_cycle=color_map_specified, properties=properties)

# make a points layer applying a color cycle to the unique properties
face_color_cycle = ['green', 'red', 'blue', 'white']
viewer.add_points(points, face_color='my_colors', face_color_cycle=color_map_specified, properties=properties)
```

**Comment by constantinpape** [comment]
Thanks for the reply @kevinyamauchi. The work-around you proposed fixes the issue for now :).

> How would you prefer the API to be? Perhaps allowing the user to pass a dictionary to the color cycle to specify the mapping and a list to allow apply a color cycle to the unique properties (as it is now)?

Yes, I think that would be a good solution!

**Comment by kevinyamauchi** [comment]
> Thanks for the reply @kevinyamauchi. The work-around you proposed fixes the issue for now :).

Great! We will keep you updated as we converge on the right way to expose this in the init method.

> Yes, I think that would be a good solution!

Thanks for the feedback!

**Comment by kevinyamauchi** [comment]
I just wanted to make a note that when we address this, we should leverage the work from @ziyangczi in #1339 (label->color mapping).

**Comment by liu-ziyang** [comment]
after reading the existing code and tried to patch in place, I start to feel we may need more drastic changes to make it more intuitive. right now there are 3 things involved in setting color as pointed out: property, face/edge_color, face/edge_color_cycle, I would propose that we do something similar to the label layer, where we consolidate them to one single dict as suggested: {1: 'green', 2: 'red', 3: 'blue', 4: 'white'} and set `face_colors`/`edge_colors` instead, and we would have the same behavior where the id would be incremental and would default to white when missing from the given map. This does mean the API would be different, so I want to confirm this is the direction we are going

**Comment by liu-ziyang** [comment]
@constantinpape out of curiosity, why would you include green in the color cycle if you wish not to see it? I am trying to understand the usecase in case I miss anything in my commit

**Comment by sofroniewn** [comment]
This is closed by #1623
